src/pybind/matrix/matrix_pybind.cc

```diff
@@ -49,12 +50,56 @@ void pybind_matrix(py::module& m) {
            [](MatrixBase<float>& m, std::pair<ssize_t, ssize_t> i, float v) {
              m(i.first, i.second) = v;
            })
-      .def("numpy", [](MatrixBase<float>* m) {
-        return py::array_t<float>(
-            {m->NumRows(), m->NumCols()},                  // shape
-            {sizeof(float) * m->Stride(), sizeof(float)},  // stride in bytes
-            m->Data(),                                     // ptr
-            py::none());  // pass a base object to avoid copy!
+      .def("numpy",
```

Comment:
For CPU kaldi matrix / pytorch tensor zero-copy, do we really need to convert to dlpack since one line of binding code,

.def_buffer([](Matrix<float> &mat) -> py::buffer_info {
      return py::buffer_info(
        mat.Data(), /*raw pointer*/
        sizeof(float),/*size of one scalar*/
        py::format_descriptor<float>::format(),
        2, /*number of dimensions*/
        {mat.NumRows(), mat.NumCols()} , /*buffer dimension*/
        {sizeof(float) * mat.Stride(), sizeof(float)} /*strides in bytes*/
      );

will accomplish zero-copy between kaldi matrix  and numpy array. And I believe all DL framework support numpy array now. therefore using numpy as a bridge instead of dlpack is more straghtforward.

Reply:
Yes, you're right.

This method `numpy` does not use DLPack interface. As far as I know, numpy does not support DLPack.

The method you mentioned allows us to convert a `kaldi::Vector<float>` to `numpy`
by using

```python
numpy_array = np.array(kaldi_vector, copy=False)
```

By adding this method, we can use

```python
numpy_array = kaldi_vector.numpy()
```
which I think is much easier and much more convenient.

Reply:
it might be more straightforward to call numpy(), but not sure it's easier/more convenient. 

I thought DLPack, is just an interface to allow you to accomplish zero-copy. Why do we need to support DLPack if we could do zero-copy w/o it?

Reply:
If I understand you correctly, you think that `DLPack` is used to convert a kaldi vector to or from numpy ndarray.

The current usage of DLPack is to zero-copy data from Kaldi to some framework, 
or zero copy-data from some framework to Kaldi, both for CPU and GPU data, as
long as the framework supports the DLPack interface.

DLPack is not used for numpy.  From C++ to Python, DLPack wraps a data ptr
and some other meta information in it and is sent as a PyCapsule to Python.
In Python, if a framework supports DLPack interface, it will consume this PyCapsule
and extracts the data ptr and the meta-information to construct a tensor specific
to this framework. 

I think DLPack has nothing to do with numpy. It is mainly used for zero-copying
data between different frameworks. Numpy does not support DLPack interface,
since we cannot construct a numpy ndarray directly from a PyCapsule.

Reply:
DLPack can also support zero-copying **GPU** data.

Reply:
I prefer
```python
numpy_array = kaldi_vector.numpy()
```

to 

```python
numpy_array = np.array(kaldi_vector, copy=False)
```

PyKaldi and PyTorch also support this kind of syntax `numpy_array = tensor.numpy()`

Reply:
I didn't mean using DLPack is to zero-copy between Kaldi vector and numpy and was aware of that DLPack is irrelevant to numpy. I was saying since all DL framework supports numpy, and with numpy using buffer protocol, we could easily zero-copying CPU data between different frameworks why bother using DLpack at all for CPU?

Reply:
I see and I agree with you.

`kaldi <--> numpy <---> some DL framework` is enough for **CPU** data
and we do NOT need `kaldi <---> DLPack <---> some DL framework`

But with DLPack, we can handle CPU/GPU data homogeneously:

For example, to zero copy data from kaldi to some DL framework, the following function is enough

```python
def from_kaldi_to_some_dl_framework(v):
  tensor = from_dlpack(v.to_dlpack())
```

Without DLPack, we have to do
```python
def from_kaldi_to_some_dl_framework(v):
    if v is on GPU:
      tensor = from_dlpack(v.to_dlpack())
    else:
      tensor = some_framework.from_numpy(v.numpy())
```

Reply:
Make much sense, this way zero-copy can be device-agostic

Reply:
😄

Reply:
Mark as resolved.